### DISCUSSION TITLE
Visibility of existing files/directories in /data during ingress

### DISCUSSION THREAD

**Discussion by DDelbarre** [alan-turing-institute/data-safe-haven#1448] [Ideas]
When ingressing data via Azure Storage Explorer, it is possible for whovever is uploading the data to see (but not read) the other directories and files in the `/data` volume of the SRE.

For projects like EDoN, where there are a number of separate data providers who will ingress data into an SRE, this could be quite problematic. This is because data providers will be able to see the files and directories uploaded by the other providers. In cases where there are files like medical images (e.g. MRI scans) these typically have patient IDs in the file names, which we would not want other providers to see.

I was wondering if it was possible to restrict what data providers can see when ingressing data? I had an idea as to how this could be implemented, but there are probably other approaches too!

We currently use the following directory structure for EDoN in `/data`:
```
2023-01-01
    file_1
    file_2
2023-02-01
    file_1
    file_2
```

If we changed to something like the following, the data providers could be given an upload token to their specific directory. That way they could still see the data that _they_ have uploaded previously (which could be useful for them to ensure they are not re-uploading duplicated data) but they would not be able to see anything else.
```
data_provider_1
    2023-01-01
        file_1
        file_2
data_provider_2
    2023-02-01
        file_1
        file_2
```

In the forms that are used for recording the details of the data ingress prior to it taking place, there could also be a section where the storage location could be specified to ensure that access is being given to the appropriate location.

**Comment by edwardchalstrey1** [comment]
I will have a proper think about this when I can, but just to quickly note that on Azure and it’s not obvious whether the token permission can set to a specific sub-folder like `/data/data_provider_1 ` for instance

**Comment by JimMadge** [comment]
The SAS token permissions are listed [here](https://learn.microsoft.com/en-us/rest/api/storageservices/create-service-sas#permissions-for-a-directory-container-or-blob).

The documentation suggests using 'list' and 'write' (**not** 'read'). So, the uploader would then be able to see the _names_ of existing blobs, but not their contents, and to write new blobs.

If we want to hide the names themselves we could remove the 'list' permission. You would need to be a bit careful to make sure that no one overwrites existing data. _E.g._ two providers upload a directory with the same name.

@DDelbarre Thanks for thinking about this. If you make this clear when making an ingress request I believe it should be possible.

@craddm @edwardchalstrey1 If either of you have a sandbox running, could you check exactly _what_ you can see with and without the list permission? _I.e._ can you see all filenames, just the top-level directories or something else.

We should be confident how this works before making an recommendations. This would be a great addition to the documentation.

**Reply by DDelbarre** [reply]
If 'list' permissions were removed for providers, presumably they would not be able see the files that they have just uploaded? I'm just wondering if this could also cause issues as the providers would not be able to check that they have ingressed all of their files (for example, if an upload was interrupted and the provider wanted to carry on from where they left off), which might lead to incomplete transfers of data.

Is there any way that we could disable overwriting of pre-existing data? If data was accidentally overwritten we probably would not be able to tell very easily, which could cause some data integrity issues too.

**Reply by edwardchalstrey1** [reply]
<img width="914" alt="Screenshot 2023-04-24 at 09 17 59" src="https://user-images.githubusercontent.com/5486164/233939513-dc04c6d9-161f-4d30-9e75-ed3a29355370.png">

@JimMadge We actually can't not have "list" and connect to the blob container via Azure Storage Explorer in the usual way for ingress ^

**Reply by JimMadge** [reply]
Ok, in that case, can you see whether we can create directories and create a SAS token with list, write _just_ for that directory?

**Reply by edwardchalstrey1** [reply]
StackOverflow answer: "You can only get Shared Access Signature for a blob container or for a single blob, but you are NOT able to get Shared Access Signature for a blob virtual directory, since directory isn't a real concept in Azure Blob Storage."

How about this for a solution:
1. I set up a new blob container for each data uploader and provide them a token for that
2. The data uploader performs the upload
3. I copy the files over to appropriate sub-folders in the ingress container with my admin access in Azure Storage Explorer, making them available to everyone logged into the SRE at `/data`, but not the uploader
4. I delete the containers that were created for this purpose

**Reply by JimMadge** [reply]
> If 'list' permissions were removed for providers, presumably they would not be able see the files that they have just uploaded? 

Yes, but even list wouldn't be sufficient to look at the uploaded data. Read permissions would be needed for that. I would be reluctant to give read permissions as it is an unnecessary privilege to upload data.

> I'm just wondering if this could also cause issues as the providers would not be able to check that they have ingressed all of their files (for example, if an upload was interrupted and the provider wanted to carry on from where they left off), which might lead to incomplete transfers of data.
> 
> Is there any way that we could disable overwriting of pre-existing data? If data was accidentally overwritten we probably would not be able to tell very easily, which could cause some data integrity issues too.

In either case, I don't think letting the data provider read input data, or preventing them from overwriting data is the solution; neither guarantee the data is correct. The data providers should really be giving checksums of the data so that it can be verified.

@edwardchalstrey1 I feel there used to be a note about this in the documentation but I can't find any reference. (I also remember at least talking about the idea of an automatic, periodic check of input data checksums). I feel we should write some guidance about this, do you think it would be helpful to have something prescriptive in the DSH docs or to let instance mangers deal with this individually?

**Reply by JimMadge** [reply]
> StackOverflow answer: "You can only get Shared Access Signature for a blob container or for a single blob, but you are NOT able to get Shared Access Signature for a blob virtual directory, since directory isn't a real concept in Azure Blob Storage."

:+1: Could you also just double check that this is true?

> How about this for a solution:
> 
>     1. I set up a new blob container for each data uploader and provide them a token for that
> 
>     2. The data uploader performs the upload
> 
>     3. I copy the files over to appropriate sub-folders in the ingress container with my admin access in Azure Storage Explorer, making them available to everyone logged into the SRE at `/data`, but not the uploader
> 
>     4. I delete the containers that were created for this purpose

@edwardchalstrey1 That sounds like a reasonable work around 🙌 , as long as the new containers have the same 'security configuration' as we would expect normally (_e.g._ restricted to specific IPs). Could you just confirm this wouldn't mean you have any more access than you normally would?

If this works, it would be good to write some guidance on supporting multiple data providers. I feel like the best thing is just to not put sensitive data in file/directory names. However, I think what you have described above @edwardchalstrey1 seems like a good way to support cases where this is difficult or a data provider feels uncomfortable about other providers seeing file/directory names.

**Reply by edwardchalstrey1** [reply]
> 👍 Could you also just double check that this is true?
Yes, there's no way to do it in Azure Portal

I think your suggestion to update the docs for multiple data providers and saying explicitly that in that case the best practice is to not have sensitive data in filenames, sounds like the best update to the docs

I propose to try out my suggestion for Edon's case, and even assuming this works as intended, not to document this as a best practice in the docs. But we can leave this discussion up here on GitHub so if future DSH System Manager's came across the exact specific problem they can still see what we did

**Reply by edwardchalstrey1** [reply]
Also just to check @DDelbarre are you happy with this? If so, I can generate these separate SAS urls for the data uploaders

**Reply by JimMadge** [reply]
> I propose to try out my suggestion for Edon's case, and even assuming this works as intended, not to document this as a best practice in the docs. But we can leave this discussion up here on GitHub so if future DSH System Manager's came across the exact specific problem they can still see what we did

I understand not wanting to recommend this as best practice. If this works I would like to record it somewhere more authoritative/findable than in a discussion thread though.

**Reply by DDelbarre** [reply]
Hi both, thanks for looking into this. I think that the proposed solution sounds good. 

> If this works, it would be good to write some guidance on supporting multiple data providers. I feel like the best thing is just to not put sensitive data in file/directory names. However, I think what you have described above @edwardchalstrey1 seems like a good way to support cases where this is difficult or a data provider feels uncomfortable about other providers seeing file/directory names.

In some cases I think it might be unavoidable to have IDs in filenames for EDoN. Going back to my example of MRI scans, the layout of the directory structure and naming of the files is controlled in these datasets using industry standard specifications (e.g. BIDS) that require the ID to be in the filenames. In other projects I have worked on we have tried zipping MRI datasets for transfer but have had issues with transfers getting interrupted due to the large file sizes and needing to restart transfers multiple times. 

Also some of our data providers in EDoN are large organisations that won't undertake custom data preparation so we don't have much control over the format of this data prior to ingress.
